### DISCUSSION TITLE
Release v3.3.0

### DISCUSSION THREAD

**Discussion by nathanlesage** [Zettlr/Zettlr#5488] [Announcements]
**Projects | Vim Mode | Quality of Life Fixes | Bug Fixes**

This update includes many bug fixes, tiny improvements, as well as full-fledged new features. Before you use it, however, please make sure to back up your writing statistics, because this update includes changes in the writing statistics subsystem that should not – but may – change your writing statistics in an unwanted way.

> **Do you enjoy Zettlr? Consider donating to help us pay for the development costs. You can donate [monthly on Patreon](https://www.patreon.com/zettlr) or [once via PayPal](https://www.paypal.me/hendrikerz). Additionally, you can [sponsor Hendrik Erz on GitHub](https://github.com/sponsors/nathanlesage). Thanks for your support!**

With that out of the way, here are the highlights. First, we have fixed a long-standing bug: Those of you who like to indent lists with tab characters instead of spaces had the problem that the list items would "jump around," or "dance" while typing. We are happy to announce that we finally found a way to stop this from happening with this update. Second, after neglecting the project feature for years, we continue our journey to offer you the best project experience by transferring your project sorting to the file manager. From now on, project folders will show you those files that you include in the project export on top of the other folder contents, and in the correct order. This way, you don't have to do any guesswork as to which file is at which position in your export order. When you change your project settings to include new or remove existing files, or change the order, this will be immediately reflected in the file manager. Note, however, that this only works with project files that are directly positioned inside the project directory. If you tried to keep your files in order using subfolders, we do recommend to "flatten out" your folder hierarchy now, as your woes should be a thing of the past with this update. Lastly, we have implemented features and improvements for users of the vim mode (thanks to steering committee member Kirthian Yasotharan for his valuable input), and the formatting toolbar can now be disabled with a setting.

Aside from these major improvements, we have included many additional smaller fixes in this update. For example, dragging and dropping and pasting files from your file explorers should work as expected again. Also, we have improved the interchangeability with other note-taking editors that support wikilinks to specific subheadings in a file, which Zettlr now finally also understands. Furthermore, images in the "Other files" tab now have a preview image, making it faster and simpler to identify images to include in your files.

To know about everything that's changed, please refer to the full changelog below.

# Changelog

## Please Back up Your Writing Statistics

This update includes a change in the writing statistics. While we extensively
tested it, we would like you to back up your personal writing statistics. You can
do so by going into Zettlr's data directory and simply make a copy of the file
`stats.json` before launching the update. Your data directory is located at
`C:\Users\<your username>\AppData\Roaming\Zettlr` (Windows),
`/Users/<your username>/Library/Application support/Zettlr` (macOS), or
`/home/<your username>/.config/Zettlr` (Linux). Then, check if your writing
statistics still look proper. If not, please report this. Thanks!

## GUI and Functionality

- **New Feature**: Images in the "other files" tab do now contain a small
  preview to help find the correct images
- **New Feature**: Project settings can now override folder sorting, which means
  that any project folder will from now on be sorted like so: (1) the files
  within the folder in the order of the project settings; (2) all other files
  according to the folder's regular sort settings. Note that this requires all
  project files to reside in the top folder; included project files in
  subfolders will not be affected
- **New Feature**: The formatting toolbar can now be toggled on or off in the
  preferences (#5207)
- **New Feature**: Allow three-digit ordinal day of the year in IDs and
  filenames (`%o`; #5424)
- **Vim mode improvements**:
  - Mapped Vim's write and quit commands to saving and closing actions (#4720,
    #5463):
    - `w`: Executes a save command for the current file
    - `q`: Executes a close-file command for the current file
    - `wq`: Attempts to save the current file and then close it
    - Note that the `!` argument for supressing the "Omit unsaved changes"
      dialog will not work, as the editor does not have the authority to tell
      main to simply omit work (this is a security feature)
  - Movement keys (`j`/`k`) now account for line wrapping for a smoother
    navigation experience
  - Default Shortcuts Restored: Unmapped `C-f`, `C-t`, and `C-c` in specific
    modes to re-enable default editor behaviors like search and task item
    shortcuts
- Columns in the preferences window are now properly aligned (#5410)
- Fix color scheme in readability mode (#5478)
- Prevent initial startup update-check if the setting is unchecked (context:
  https://github.com/Zettlr/Zettlr/commit/812899#r148519528)
- Fix errors in, and update, German translation (`de-DE`)
- Update Turkish translation (`tr-TR`) (#5461)
- Fix "dancing list items" (#4602)
- Zettlr finally counts and remembers character counts
- The link preview and force-open workflows now expect internal/wiki-links that
  link to headings within the file (`[[filename#heading-id]]`) and can handle
  such links appropriately; the editors will not yet jump to the corresponding
  headings, but this will improve compatibility with other editors who already
  support this feature (#3727)
- Fixed an edge case in rendering highlight marks
- Fixed non-unique clipboard data paste filenames (#5449)
- Fixed non-proportional images (especially very tall ones) overlaying other
  text below them (#5465)
- The background color of the active line in typewriter mode no longer blocks
  the selection background (#5430)
- Dropping or copy-pasting images from the file browser now inserts relative
  links to them again instead of offering to save a copy (#5475)
- Code files can now also be indented and unindented as expected using `Tab`
- Changing capitalization of filenames on Windows, macOS, or other case-
  insensitive file systems (e.g., testfile -> Testfile) no longer fails (#5460)
- Fixed drag & drop behavior of open-able files onto the editor from the file
  browser; dropping any supported file onto the app will now attempt to open it
  (#5344)
- Fixed export menu not remembering last selected custom commands (#5163)
- Provide default reset-to-sizes for various split views

## Under the Hood

- Update Codemirror dependencies
- Update Electron to `v33.2.0`
- Update Electron forge to `v7.5.0`
- Update various other dependencies
- The `PersistentDataContainer` now uses proper data types, making usage more
  type-safe
- Completely refactor the Statistics provider as well as the statistics display
- Replace the deprecated `registerFileProtocol` call with the new recommended
  `handle` call for handling `safe-file://` calls
- Re-introduce linter rules; all of these are part of the common style we
  already use, but they got lost in one of the past ESLint upgrades:
  - enforce single quotes across the codebase
  - enforce proper object property spacing
  - enforce type imports
- Moved all keymaps to a centralized space, enabling us to customize them
  further in the future and streamlining the available keybindings
- The document provider is now more strict when loading persisted window
  arrangements from disk on startup
- Made many additional strings in the GUI translatable

## New Contributors
* @Onatcer made their first contribution in https://github.com/Zettlr/Zettlr/pull/5451
* @EnginKarahan made their first contribution in https://github.com/Zettlr/Zettlr/pull/5455
* @hnsstrk made their first contribution in https://github.com/Zettlr/Zettlr/pull/5478

**Full Changelog**: https://github.com/Zettlr/Zettlr/compare/v3.2.3...v3.3.0

<hr /><em>This discussion was created from the release <a href='https://github.com/Zettlr/Zettlr/releases/tag/v3.3.0'>Release v3.3.0</a>.</em>

**Comment by marziod** [comment]
Hello. FYI: i updated from a previous installation, and now images are broken , and even if i click on external links, they does not work!

**Reply by kyaso** [reply]
Please open an issue for this here on GitHub providing additional details to help us fix the problem as fast as possible.

**Comment by BabakFarshchian** [comment]
@nathanlesage I tried to send you a payment in PayPal but it got returned :-)

**Reply by nathanlesage** [reply]
Did you send it as a gift or as a payment for services? The latter I think often gets returned, so make sure to select the former option. But thanks anyways, appreciate it!
